src/klein/test/test_headers.py

```diff
@@ -0,0 +1,548 @@
+    def test_headerValueAsBytesWithBytes(self, value):
+        # type: (bytes) -> None
+        """
+        L{headerValueAsBytes} passes through L{bytes}.
+        """
+        self.assertIdentical(headerValueAsBytes(value), value)
+
+
+    @given(text(min_size=1))
+    def test_headerValueAsBytesWithText(self, value):
+        # type: (Text) -> None
+        """
+        L{headerValueAsBytes} encodes L{Text} using L{HEADER_VALUE_ENCODING}.
+        """
+        rawValue = encodeName(value)
+        assume(rawValue is not None)
+        self.assertEqual(headerValueAsBytes(value), rawValue)
+
+
+    @given(binary())
+    def test_headerValueAsTextWithBytes(self, value):
+        # type: (bytes) -> None
+        """
+        L{headerValueAsText} decodes L{bytes} using L{HEADER_VALUE_ENCODING}.
+        """
+        self.assertEqual(headerValueAsText(value), decodeValue(value))
+
+
+    @given(text(min_size=1))
+    def test_headerValueAsTextWithText(self, value):
+        # type: (Text) -> None
+        """
+        L{headerValueAsText} passes through L{Text}.
+        """
+        self.assertIdentical(headerValueAsText(value), value)
+
+
+
+class HeaderNameNormalizationTests(TestCase):
+    """
+    Tests for header name normalization.
+    """
+
+    def test_normalizeLowerCase(self):
+        # type: () -> None
+        """
+        L{normalizeHeaderName} normalizes header names to lower case.
+        """
+        self.assertEqual(normalizeHeaderName("FooBar"), "foobar")
+
+
+
+class RawHeadersConversionTests(TestCase):
+    """
+    Tests for L{normalizeRawHeaders}.
+    """
+
+    def test_pairNotTuple(self):
+        # type: () -> None
+        """
+        L{normalizeRawHeaders} raises L{TypeError} if the C{headerPairs}
+        argument is not an tuple L{tuple}s.
+        """
+        e = self.assertRaises(
+            TypeError, tuple, normalizeRawHeaders(([b"k", b"v"],))
+        )
+        self.assertEqual(str(e), "header pair must be a tuple")
+
+
+    def test_pairsWrongLength(self):
+        # type: () -> None
+        """
+        L{normalizeRawHeaders} raises L{ValueError} if the C{headerPairs}
+        argument is not an tuple of 2-item L{tuple}s.
+        """
+        for pair in ((b"k",), (b"k", b"v", b"x")):
+            e = self.assertRaises(
+                ValueError,
+                tuple,
+                normalizeRawHeaders(cast(Iterable[Iterable[bytes]], (pair,))),
+            )
+            self.assertEqual(str(e), "header pair must be a 2-tuple")
+
+
+    def test_pairsNameNotBytes(self):
+        # type: () -> None
+        """
+        L{normalizeRawHeaders} raises L{TypeError} if the C{headerPairs}
+        argument is not an tuple of 2-item L{tuple}s where the first item in
+        the 2-item L{tuple} is L{bytes}.
+        """
+        e = self.assertRaises(
+            TypeError,
+            tuple,
+            normalizeRawHeaders(
+                cast(Iterable[Iterable[bytes]], ((u"k", b"v"),))
```

Comment:
OK, what's going on here is that `normalizeRawHeaders` expects an iterable of iterables of `bytes`, but the data we are passing in here is an iterable of iterables of `object`, because we're intentionally passing in the wrong thing in the test to make sure we raise an exception.

That is: the type checker, if you are using it, catches the incorrect usage (hence the need for the cast), but the runtime test also catches it.